### PR TITLE
[node-core-library] Implement json-schema 'draft-07' validation

### PR DESCRIPTION
**Pull request by bartvandenende-wm**
## Summary

[node-core-library] Implements json-schema 'draft-07' validation in JsonFile / JsonSchema

Fixes https://github.com/microsoft/rushstack/issues/4713

## Details

This PR:
* Upgrades 'ajv' from v6.x to v8.x in the rush repo 
  * see for more information https://ajv.js.org/v6-to-v8-migration.html
* Replaces the 'z-schema' library with 'ajv' in `@rushstack/node-core-library`
  * This migration is required to be able to properly support json-schema draft-07 (and other future specifications), z-schema does not see very active development anymore.
  * Updates the schema validator error-detail helper to error details exposed by ajv
    * Since ajv uses a slightly different error message format, there are a couple of small breaking changes in downstream project that have test assertions on them.
  * Use 'ajv-draft-04' to retain backwards compatibility with json-schema draft-04
    * see https://ajv.js.org/guide/schema-language.html#draft-04
  * `ajv` applies additional validation on the correctness of the schema per https://ajv.js.org/options.html#strictschema
    * this surfaced a small issue in the heft schema validation where 'items' keyword was incorrectly combined with the `string` type.

Notes: 
- Schema definitions that don't have a schema defined will now default to draft-07.
- There are likely some performance gains with the migration from 'z-schema' to 'ajv' that where not quantified, more background https://ajv.js.org/guide/why-ajv.html#super-fast-secure.

## How it was tested

1. validated the existing unit tests and updated the new error detail snapshot
2. added new unit test to specifically test draft-04, draft-07 and nested schemas

## Impacted documentation

N/A

### CODE REVIEW

File: common/changes/@rushstack/rig-package/bartvandenende-wm-support-json-schema-draft-07_2024-05-15-14-55.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@rushstack/rig-package",
+      "comment": "Update build dependencies",
+      "type": "patch"
```
Comment (iclanton) [inline common/changes/@rushstack/rig-package/bartvandenende-wm-support-json-schema-draft-07_2024-05-15-14-55.json:6]:
```suggestion
      "comment": "",
      "type": "none"
```

Changes to `devDependencies` aren't relevant to a consumer of the package.

File: common/changes/@microsoft/rush/bartvandenende-wm-support-json-schema-draft-07_2024-05-15-16-44.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Update test assertions for schema validation failures.",
+      "type": "none"
```
Comment (iclanton) [inline common/changes/@microsoft/rush/bartvandenende-wm-support-json-schema-draft-07_2024-05-15-16-44.json:6]:
```suggestion
      "comment": "",
      "type": "none"
```

This isn't relevant to a consumer of the package.

File: libraries/node-core-library/src/JsonSchema.ts [hunk 3]
```diff
@@ -210,6 +215,18 @@ export class JsonSchema {
...
+   * Helper function to determine if the targeted schema is using the draft-04 specification
+   * @returns boolean value indicating that the schema uses the draft-04 specification
+   */
+  private _isDraft04Schema(): boolean {
+    if (this._schemaObject.$schema) {
+      return this._schemaObject.$schema.startsWith('http://json-schema.org/draft-04/');
+    } else {
+      return false;
+    }
+  }
```
Comment (iclanton) [inline libraries/node-core-library/src/JsonSchema.ts:228]: If someone doesn't have a `$schema` property, we're defaulting to draft-07?

Can we include an option to the initializers' `options` object to explicitly specify which version of the schema the 
consumer wants?

File: libraries/node-core-library/src/JsonSchema.ts [hunk 4]
```diff
@@ -242,16 +263,16 @@ export class JsonSchema {
       // Validate each schema in order.  We specifically do not supply them all together, because we want
       // to make sure that circular references will fail to validate.
       for (const collectedSchema of collectedSchemas) {
-        if (!newValidator.validateSchema(collectedSchema._schemaObject)) {
+        if (!validator.addSchema(collectedSchema._schemaObject)) {
           throw new Error(
             `Failed to validate schema "${collectedSchema.shortName}":` +
               os.EOL +
-              JsonSchema._formatErrorDetails(newValidator.getLastErrors())
+              JsonSchema._formatErrorDetails(validator.errors!)
```
Comment (iclanton) [inline libraries/node-core-library/src/JsonSchema.ts:270]: We know `errors` is defined in this case?

File: libraries/heft-config-file/src/test/__snapshots__/ConfigurationFile.test.ts.snap [hunk 5]
```diff
@@ -272,12 +272,12 @@ exports[`ConfigurationFile error cases Throws an error when a combined config fi
 "Resolved configuration object does not match schema: Error: JSON validation failed:
 <project root>/lib/test/errorCases/invalidCombinedFile/config1.json
 
-Error: #/
-       Data does not match any schemas from 'oneOf'
-  Error: #/
-         Additional properties not allowed: folderPaths
-  Error: #/
-         Additional properties not allowed: filePaths"
+Error: # (must NOT have additional properties)
```
Comment (iclanton) [inline libraries/heft-config-file/src/test/__snapshots__/ConfigurationFile.test.ts.snap:275]: The parenthesis on these lines aren't particularly useful.

File: libraries/node-core-library/src/test/JsonSchema.test.ts [hunk 6]
```diff
@@ -8,25 +8,84 @@ describe(JsonSchema.name, () => {
   const schemaPath: string = `${__dirname}/test-data/test-schema.json`;
   const schema: JsonSchema = JsonSchema.fromFile(schemaPath);
 
-  test('loadAndValidate successfully validates a JSON file', () => {
-    const jsonPath: string = `${__dirname}/test-data/test.json`;
-    const jsonObject: JsonObject = JsonFile.loadAndValidate(jsonPath, schema);
+  describe('loadAndValidate', () => {
```
Comment (iclanton) [inline libraries/node-core-library/src/test/JsonSchema.test.ts:11]:
```suggestion
  describe(JsonFile.loadAndValidate.name, () => {
```

**Review by iclanton — COMMENTED** [review]
Looks good overall.

I'm going to disable automatic publishing for now. If we're going to do a major bump of `node-core-library`, there are some other changes we should make at the same time.

File: libraries/node-core-library/src/JsonSchema.ts (hunk 3 above)
Comment (bartvandenende-wm) [reply libraries/node-core-library/src/JsonSchema.ts:228]: yes the `ajv` defaults to the json-schema version that is linked to the provided 'ajv' module export, which defaults  to draft-07. There are alternative module exports available for draft-04, draft-2019 and draft-2020.

We can add an option to force a specific json-schema version, but the downside is that we would be introducing a secondary mechanism to provide a schema version next to the schema itself. We could also see conflicts if mismatched versions are provided, so I'm not sure if we gain a lot by doing so...

File: libraries/heft-config-file/src/test/__snapshots__/ConfigurationFile.test.ts.snap (hunk 5 above)
Comment (bartvandenende-wm) [reply libraries/heft-config-file/src/test/__snapshots__/ConfigurationFile.test.ts.snap]: yeah that is fair, ajv does not have 'detailed' error object like z-schema and present all relevant info in the error 'message'. so we can probably remove the parenthesis section completely

File: libraries/node-core-library/src/JsonSchema.ts (hunk 4 above)
Comment (bartvandenende-wm) [reply libraries/node-core-library/src/JsonSchema.ts]: yes however I noticed that I failed to notice that `ajv.addSchema` has a different return type as `ajv.validateSchema` (ajv.addSchema uses ajv.validateSchema under the hood). I just made a tweak in the condition statement to correctly assert the error here, which also remove the needs to add the exclamation mark here.

File: libraries/node-core-library/src/JsonSchema.ts (hunk 3 above)
Comment (iclanton) [reply libraries/node-core-library/src/JsonSchema.ts:218]: I think it'd be useful to be able to explicitly dictate which version of the schema you want to target. It'd also be useful to create an API to dictate the version (either with an enum or a string union) to create precedent for supporting future versions.

Comment (bartvandenende-wm) [reply libraries/node-core-library/src/JsonSchema.ts]: 👍 I provided an initial version of expose a schema version option, let me know what you think

Comment (iclanton) [reply libraries/node-core-library/src/JsonSchema.ts]: Pushed a few tweaks.

**Review by iclanton — APPROVED** [review]

**Pull request merged**